File: backend/src/onedrop/api/routers/events.py
"""Event endpoints."""

from __future__ import annotations

import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query, Response, status

from onedrop.api.dependencies import CurrentUser, DbSession
from onedrop.db.repositories.events import MAX_PAGE_SIZE
from onedrop.events.schemas import (
    EventCreate,
    EventPage,
    EventResponse,
    EventUpdate,
    EventWithConflicts,
)
from onedrop.events.service import EventService

router = APIRouter(prefix="/events", tags=["events"])


@router.get("", response_model=EventPage)
async def list_events(
    session: DbSession,
    user: CurrentUser,
    period: Annotated[str, Query(pattern="^(day|week)$")] = "day",
    anchor_date: Annotated[date | None, Query(alias="date")] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_PAGE_SIZE)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> EventPage:
    """Day or week agenda in the user's local calendar."""
    items, has_more, resolved_date = await EventService(session).list_period(
        user.id, period=period, anchor=anchor_date, limit=limit, offset=offset
    )
    return EventPage(
        items=[EventResponse.model_validate(item) for item in items],
        range="week" if period == "week" else "day",
        anchor_date=resolved_date,
        limit=limit,
        offset=offset,
        has_more=has_more,
    )


@router.post("", response_model=EventWithConflicts, status_code=status.HTTP_201_CREATED)
async def create_event(
    payload: EventCreate, session: DbSession, user: CurrentUser
) -> EventWithConflicts:
    """Create an event. Overlapping events are reported, not blocked."""
    event, conflicts = await EventService(session).create(user.id, payload)
    return EventWithConflicts(
        event=EventResponse.model_validate(event), conflicts=conflicts
    )


@router.get("/{event_id}", response_model=EventResponse)
async def get_event(
    event_id: uuid.UUID, session: DbSession, user: CurrentUser
) -> EventResponse:
    event = await EventService(session).get(user.id, event_id)
    return EventResponse.model_validate(event)


@router.patch("/{event_id}", response_model=EventWithConflicts)
async def patch_event(
    event_id: uuid.UUID, payload: EventUpdate, session: DbSession, user: CurrentUser
) -> EventWithConflicts:
    event, conflicts = await EventService(session).patch(user.id, event_id, payload)
    return EventWithConflicts(
        event=EventResponse.model_validate(event), conflicts=conflicts
    )


@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_event(
    event_id: uuid.UUID, session: DbSession, user: CurrentUser
) -> Response:
    await EventService(session).delete(user.id, event_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/lib/events.ts
import { api } from '@/lib/api';

export const EVENT_PERIODS = ['day', 'week'] as const;
export type EventPeriod = (typeof EVENT_PERIODS)[number];
export const EVENT_PAGE_SIZE = 50;

export interface EventRecord {
  id: string;
  title: string;
  starts_at: string;
  ends_at: string | null;
  location: string | null;
  description: string | null;
  status: string;
  source_inbox_item_id: string | null;
  created_at: string;
}

export interface EventPage {
  items: EventRecord[];
  range: EventPeriod;
  anchor_date: string;
  limit: number;
  offset: number;
  has_more: boolean;
}

export interface EventWithConflicts {
  event: EventRecord;
  conflicts: string[];
}

export interface EventDraft {
  title: string;
  starts_at: string;
  ends_at?: string | null;
  location?: string | null;
  description?: string | null;
}

export interface EventChanges {
  title?: string;
  starts_at?: string;
  ends_at?: string | null;
  location?: string | null;
  status?: 'planned' | 'done' | 'cancelled';
}

export function eventListKey(period: EventPeriod, anchor: string): string[] {
  return ['events', period, anchor];
}

export function eventsPath(period: EventPeriod, anchor: string): string {
  const query = new URLSearchParams({ period, limit: String(EVENT_PAGE_SIZE) });
  if (anchor) query.set('date', anchor);
  return `/events?${query.toString()}`;
}

export function listEvents(period: EventPeriod, anchor: string): Promise<EventPage> {
  return api.get<EventPage>(eventsPath(period, anchor));
}

export function createEvent(draft: EventDraft): Promise<EventWithConflicts> {
  return api.post<EventWithConflicts>('/events', draft);
}

export function patchEvent(id: string, changes: EventChanges): Promise<EventWithConflicts> {
  return api.patch<EventWithConflicts>(`/events/${id}`, changes);
}

export function deleteEvent(id: string): Promise<void> {
  return api.del<void>(`/events/${id}`);
}
